src/ansi-c/parser.y

```diff
@@ -3407,6 +3407,15 @@ cprover_function_contract_sequence_opt:
           /* nothing */
           { init($$); }
         | cprover_function_contract_sequence
+        {
+          // Function contracts should either be attached to a
+          // top-level function declaration or top-level function
+          // definition.  Any embedded function pointer scopes should
+          // be disallowed.
+          int contract_in_global_scope = (PARSER.scopes.size() == 1);
+          int contract_in_top_level_function_scope = (PARSER.scopes.size() == 2);
+          assert(contract_in_global_scope || contract_in_top_level_function_scope);
```

Comment:
What happens if you say something like 
yyerror("function contracts allowed only at top-level declarations.")

Reply:
@jimgrundy Is this what you had in mind?
```
          int contract_in_global_scope = (PARSER.scopes.size() == 1);
          int contract_in_top_level_function_scope = (PARSER.scopes.size() == 2);
          if(!contract_in_global_scope && !contract_in_top_level_function_scope)
          {
            yyerror("Function contracts allowed only at top-level declarations.");
          }
          assert(contract_in_global_scope || contract_in_top_level_function_scope);
```

It results in an additional line of output.
```
Parsing main.c
file main.c line 8 function foo: Function contracts allowed only at top-level declarations. before ')'
Assertion failed: (contract_in_global_scope || contract_in_top_level_function_scope), function yyansi_cparse, file parser.y, line 3421.
Abort trap: 6
```

Reply:
Almost. I think we should do that to get a better error message and then error out more gracefully than an assertion violation. I don't think there should be any way user input should cause an internal assertion to fail -- failure of internal assertions should be seen as evidence of a tool bug.

Parsers don't exit immediately on the first error either.  How about this instead, this is the function the lexer uses to report errors, but it is also available in the parser:
yyansi_cerror("Function contracts allowed only at top-level declarations.").

Maybe instead of the assertion violation we just set $$ to something empty so the bad thing never gets declared? There should be some way to keep going for now but record the fact the the parse isn't clean, maybe yyansi_cerror will do that?

Reply:
@jimgrundy I've updated the error handling to print debug information through `yyerror` and then about with `YYABORT`.  This produces output like:
```
CBMC version 5.63.0 (cbmc-5.63.0-29-g6a8c9a673-dirty) 64-bit x86_64 macos
Parsing ../embedded_contract_fail_02/main.c
file ../embedded_contract_fail_02/main.c line 8 function foo: Function contracts allowed only at top-level declarations. before ')'
PARSING ERROR
```